params/precompile_config.go

```diff
@@ -61,153 +71,135 @@ func (p *PrecompileUpgrade) getByAddress(address common.Address) (precompile.Sta
-		for i, upgrade := range c.PrecompileUpgrades {
-			config, ok := upgrade.getByAddress(address)
-			// Skip the upgrade if it's not relevant to [address].
-			if !ok {
-				continue
-			}
+		upgradeTimestamp := upgrade.Timestamp()
 
-			if disabled == config.IsDisabled() {
-				return fmt.Errorf("PrecompileUpgrades[%d] disable should be [%v]", i, !disabled)
-			}
-			if lastUpgraded != nil && (config.Timestamp().Cmp(lastUpgraded) <= 0) {
-				return fmt.Errorf("PrecompileUpgrades[%d] config timestamp (%v) <= previous timestamp (%v)", i, config.Timestamp(), lastUpgraded)
-			}
+		if upgradeTimestamp == nil {
+			return fmt.Errorf("PrecompileUpgrade (%s) at [%d]: block timestamp cannot be nil ", key, i)
+		}
+		// Verify specified timestamps are monotonically increasing across all precompile keys.
+		// Note: It is OK for multiple configs of DIFFERENT keys to specify the same timestamp.
+		if previousUpgradeTimestamp != nil && upgradeTimestamp.Cmp(previousUpgradeTimestamp) < 0 {
+			return fmt.Errorf("PrecompileUpgrade (%s) at [%d]: config block timestamp (%v) < previous timestamp (%v)", key, i, upgradeTimestamp, previousUpgradeTimestamp)
+		}
 
-			if err := config.Verify(); err != nil {
-				return err
-			}
+		if disabled == upgrade.IsDisabled() {
+			return fmt.Errorf("PrecompileUpgrade (%s) at [%d]: disable should be [%v]", key, i, !disabled)
+		}
+		// Verify specified timestamps are monotonically increasing across same precompile keys.
+		// Note: It is NOT OK for multiple configs of the SAME key to specify the same timestamp.
+		if lastTimestamp != nil && (upgradeTimestamp.Cmp(lastTimestamp) <= 0) {
+			return fmt.Errorf("PrecompileUpgrade (%s) at [%d]: config block timestamp (%v) <= previous timestamp of same key (%v)", key, i, upgradeTimestamp, lastTimestamp)
+		}
 
-			disabled = config.IsDisabled()
-			lastUpgraded = config.Timestamp()
+		if err := upgrade.Verify(); err != nil {
+			return err
+		}
+
+		lastPrecompileUpgrades[key] = lastUpgradeData{
+			disabled:       upgrade.IsDisabled(),
+			blockTimestamp: upgradeTimestamp,
 		}
+
+		previousUpgradeTimestamp = upgradeTimestamp
 	}
 
 	return nil
 }
 
-// getActivePrecompileConfig returns the most recent precompile config corresponding to [address].
+// GetActivePrecompileConfig returns the most recent precompile config corresponding to [address].
 // If none have occurred, returns nil.
-func (c *ChainConfig) getActivePrecompileConfig(blockTimestamp *big.Int, address common.Address, upgrades []PrecompileUpgrade) precompile.StatefulPrecompileConfig {
-	configs := c.getActivatingPrecompileConfigs(nil, blockTimestamp, address, upgrades)
+func (c *ChainConfig) GetActivePrecompileConfig(address common.Address, blockTimestamp *big.Int) precompile.StatefulPrecompileConfig {
+	configs := c.getActivatingPrecompileConfigs(address, nil, blockTimestamp, c.PrecompileUpgrades)
 	if len(configs) == 0 {
 		return nil
 	}
 	return configs[len(configs)-1] // return the most recent config
```

Comment:
I wonder if there is a bit of room for more efficiency here. We can also look as a followup task.